odoo/tools/profiler.py

```diff
@@ -150,10 +150,25 @@ def start(self):
         init_thread = self.profiler.init_thread
         if not hasattr(init_thread, 'query_hooks'):
             init_thread.query_hooks = []
+            init_thread.pre_query_hooks = []
         init_thread.query_hooks.append(self.hook)
+        if self.profiler.params.get('explain_sql'):
+            init_thread.pre_query_hooks.append(self.pre_hook)
 
     def stop(self):
         self.profiler.init_thread.query_hooks.remove(self.hook)
+        if self.profiler.params.get('explain_sql'):
+            self.profiler.init_thread.pre_query_hooks.remove(self.pre_hook)
+
+    def pre_hook(self, cr, query, params):
```

Comment:
I'm not sure that adding that In another entry is a good idea. And the temporal information on the analyze entry doesn't really makes sense I think. This will double the sql entries and we need to filter them for other processing. 

Since for a instance of a collector, we will enable analyze for all of them, we could just store the "last analyze" on the collector maybe and store it beside the query on next `hook`. It is sad that we cannot make that in the hook actually. 

The only special case I have in mind is if the query fails (not a big deal), but we could actually move the query increment and hooks in the finally.

Reply:
This comment was on the `pre_hook`, it looks like github messed where I added this comment

Reply:
> I'm not sure that adding that In another entry is a good idea.

What are you suggesting?

> And the temporal information on the analyze entry doesn't really makes sense I think. This will double the sql entries and we need to filter them for other processing.

Without the time, it is hard to find the query because it would be tiny on the stack. It actually makes sense to see it next to the query as it is what is happening in reality.

> Since for a instance of a collector, we will enable analyze for all of them, we could just store the "last analyze" on the collector maybe and store it beside the query on next `hook`

Not sure I follow. You want to store the last query and run the EXPLAIN afterwards in the next `hook`? But then the plan would not be linked to the correct query!? 

> Another solution would be to have a specific collector for that.

I actually started with a specific collector but I don't see what it changes. AFAIK, it would only make it possible to stack the query plan and the query, but it is less correct to display it like that.

Reply:
> Not sure I follow. You want to store the last query and run the EXPLAIN afterwards in the next hook? But then the plan would not be linked to the correct query!?

The pre hook executes the explain and stores it on the collector instead of adding a new entry. 

The hook calls `self.add` with an extra data: the last explain. 

This way displaying of not the explain becomes a speed scope generation choice and not a constraint. from the profiling. All the speed scope api to define custom outputs works and it would be easy to add this option.

We expect each entry of the sql collector to be an executed query. It is great to be able to analyse but it is a data linked to the query, not an additional one :) 

> Without the time, it is hard to find the query because it would be tiny on the stack. It actually makes sense to see it next to the query as it is what is happening in reality.

Displaying it in speedscopes looks more like a convenience than what we want. It won't be readable. It is hard to distinct between analyze and actual queries. 

We could maybe try to display that a smarter way but I think that it would be to difficult to do in the speed scope view.